### PR TITLE
Port `throttle` operator to new executor

### PR DESCRIPTION
**Pull request by jachris**
## 🔍 Problem

- `throttle` still runs only on the legacy `crtp_operator` executor.

## 🛠️ Solution

- Add a new `Throttle` class deriving from `Operator<table_slice, table_slice>`.
- Reuse the old window / cutoff / drop semantics; swap `ctrl.set_waiting` + `run_delayed_weak` for `co_await sleep_until(...)`.
- Plugin now also implements `OperatorPlugin::describe()` with `Describer` + `validate()`; the legacy path is kept for backward compatibility.
- `snapshot()` serializes `total_`; `start_` is intentionally dropped as a non-portable `steady_clock` value.
- Drop the custom `tenzir.metrics.throttle` metric on the new path (consistent with `buffer`, pending generic counter/gauge support).
- Migrate the 5 existing tests from `test-legacy/` to `test/`; add error-path tests for `rate=0` and `window=0s`.

## 💬 Review

- Please double-check the drop-path corner case around `total_` reset semantics after a partial-slice drop — behavior intentionally matches the old implementation (see `find_cutoffs`), but it means that if more slices arrive inside the same window after a partial drop, some events can pass before the next rate check.
- The drop of the custom metric schema is a known behavioral loss; aligning with the TODO in `buffer.cpp`.

Resolves TNZ-117.

### CODE REVIEW

File: libtenzir/builtins/operators/throttle.cpp [hunk 1]
```diff
@@ -222,7 +226,150 @@ class throttle_operator final : public crtp_operator<throttle_operator> {
...
+      if (auto uints = part.as<uint64_type>()) {
+        for (const auto weight : uints->values()) {
+          offset += 1;
+          if (is_cutoff(weight)) {
+            co_yield offset;
+            total_ = 0;
+          }
+        }
+        continue;
+      }
```
Comment (tobim) [inline libtenzir/builtins/operators/throttle.cpp:357]: This can surely be written in a way that doesn't duplicate the logic between int64 und uint64?

File: libtenzir/builtins/operators/throttle.cpp [hunk 2]
```diff
@@ -222,7 +226,150 @@ class throttle_operator final : public crtp_operator<throttle_operator> {
...
+        co_return;
+      }
+      co_await sleep_until(*start_ + args_.window.inner);
+      start_ = std::chrono::steady_clock::now();
+      total_ = 0;
+    }
+    if (args_.drop) {
+      // Find the first cutoff, if any, and drop everything after it. The
+      // generator resets `total_` to 0 when yielding the cutoff, matching the
+      // semantics of the old implementation.
```
Comment (tobim) [inline libtenzir/builtins/operators/throttle.cpp:266]: The second sentence should not be phrased like this. We don't want to refer to another implementation here. The comment is also wrong, because the find_cutoffs loop breaks after the first yield, so the `total_ = 0` assignment is never reached.

File: libtenzir/builtins/operators/throttle.cpp [hunk 3]
```diff
@@ -222,7 +226,150 @@ class throttle_operator final : public crtp_operator<throttle_operator> {
...
+    // restarts. Leaving it out means a fresh window starts on restore, which
+    // can briefly admit more events than the configured rate. We accept this
+    // over the complexity of reconstructing wall-clock anchored windows.
+    serde("total_", total_);
+  }
+
+private:
+  auto find_cutoffs(const table_slice& slice, diagnostic_handler& dh)
+    -> generator<size_t> {
+    const auto weights = eval(args_.weight, slice, dh);
```
Comment (tobim) [inline libtenzir/builtins/operators/throttle.cpp:308]: Very optional: Maybe we want a fast path in case the expression is constant?

**Review by tobim — APPROVED** [review]
Good to have this as well now.

File: libtenzir/builtins/operators/throttle.cpp [hunk 4]
```diff
@@ -222,7 +226,150 @@ class throttle_operator final : public crtp_operator<throttle_operator> {
...
+    }
+    if (now - *start_ >= args_.window.inner) {
+      start_ = now;
+      total_ = 0;
+    }
+    // Preemptive check: the previous slice already exhausted the window budget.
+    if (total_ >= args_.rate.inner) {
+      if (args_.drop) {
+        diagnostic::warning("dropping input due to rate limit")
+          .primary(*args_.drop)
```
[the remark this answers is not in the document]
Comment (jachris) [reply libtenzir/builtins/operators/throttle.cpp:255]: 🤖 Leaving this unchanged: the new executor no longer passes empty slices around, so throttle does not need to preserve the legacy rows == 0 handling here.

File: libtenzir/builtins/operators/throttle.cpp [hunk 5]
```diff
@@ -222,7 +226,150 @@ class throttle_operator final : public crtp_operator<throttle_operator> {
...
+      co_await push(subslice(input, begin, input.rows()));
+    }
+  }
+
+  auto snapshot(Serde& serde) -> void override {
+    // `start_` is a `steady_clock::time_point` and is not portable across
+    // restarts. Leaving it out means a fresh window starts on restore, which
+    // can briefly admit more events than the configured rate. We accept this
+    // over the complexity of reconstructing wall-clock anchored windows.
+    serde("total_", total_);
```
[the remark this answers is not in the document]
Comment (jachris) [reply libtenzir/builtins/operators/throttle.cpp]: 🤖 Fixed in ab829f64585 by dropping the snapshot override, so we no longer restore total_ without a matching window anchor.

File: libtenzir/builtins/operators/throttle.cpp (hunk 1 above)
Comment (jachris) [reply libtenzir/builtins/operators/throttle.cpp:348]: 🤖 Fixed in ab829f64585 by sharing the cutoff iteration logic between signed and unsigned weight series.

File: libtenzir/builtins/operators/throttle.cpp (hunk 2 above)
Comment (jachris) [reply libtenzir/builtins/operators/throttle.cpp]: 🤖 Fixed in ab829f64585 by removing the stale sentence and keeping the comment scoped to the drop-path behavior.

File: libtenzir/builtins/operators/throttle.cpp (hunk 3 above)
Comment (jachris) [reply libtenzir/builtins/operators/throttle.cpp:298]: 🤖 Leaving this out for now. The same constant-expression evaluation overhead exists in a lot of other operators, so I would rather handle it as a broader eval or planning optimization than special-case throttle here.
